applications/FluidDynamicsApplication/custom_elements/data_containers/qs_vms/qs_vms_residual_derivatives.h

```diff
@@ -0,0 +1,562 @@
+
+            // calculate shear stress derivative.
+            const auto& r_strain_rate_variables = QSVMSDerivativeUtilities<TDim>::GetStrainRateVariables();
+            Vector value;
+            rData.mShearStressDerivative.clear();
+
+            // calculate shear stress derivatives w.r.t. strain rate
+            for (IndexType i = 0; i < TStrainSize; ++i) {
+                rData.mpConstitutiveLaw->CalculateDerivative(rData.mConstitutiveLawValues, CAUCHY_STRESS_VECTOR, *r_strain_rate_variables[i], value);
+                noalias(rData.mShearStressDerivative) += value * rData.mStrainRateDerivative[i];
+            }
+            // calculate shear stress derivatives w.r.t. effective viscosity
+            rData.mpConstitutiveLaw->CalculateDerivative(rData.mConstitutiveLawValues, CAUCHY_STRESS_VECTOR, EFFECTIVE_VISCOSITY, value);
+            noalias(rData.mShearStressDerivative) += value * effective_viscosity_derivative;
+
+            AddViscousDerivative(rData, rResidualDerivative, NodeIndex,
+                                       W, rN, rdNdX, WDerivative,
+                                       DetJDerivative, rdNdXDerivative);
+        }
+
+        ///@}
+
+    private:
+        ///@name Private members
+        ///@{
+
+        ResidualsContributions mResidualsContributions;
+
+        ///@}
+        ///@name Private Operations
+        ///@{
+
+        void static AddViscousDerivative(
+            Data& rData,
+            VectorF& rResidualDerivative,
+            const int NodeIndex,
+            const double W,
+            const Vector& rN,
+            const Matrix& rdNdX,
+            const double WDerivative,
+            const double DetJDerivative,
+            const Matrix& rdNdXDerivative)
+        {
+            BoundedMatrix<double, TStrainSize, TElementLocalSize> strain_matrix_derivative;
+            FluidElementUtilities<TNumNodes>::GetStrainMatrix(rdNdXDerivative, strain_matrix_derivative);
+
+            const VectorF& rhs_contribution_derivative =
+                prod(trans(rData.mStrainMatrix), rData.mShearStressDerivative) +
+                prod(trans(strain_matrix_derivative), rData.mShearStress);
+
+            noalias(rResidualDerivative) -= rhs_contribution_derivative * W;
+        }
+
+        ///@}
+    };
+
+    template<unsigned int TComponentIndex>
+    class SecondDerivatives
+    {
+    public:
+        ///@name Type definitions
+        ///@{
+
+        constexpr static IndexType NumNodes = TNumNodes;
+
+        constexpr static IndexType BlockSize = TBlockSize;
+
+        ///@}
+        ///@name Operations
+        ///@{
+
+        void CalculateGaussPointResidualsDerivativeContributions(
+            VectorF& rResidualDerivative,
+            Data& rData,
+            const int NodeIndex,
+            const double W,
+            const Vector& rN,
+            const Matrix& rdNdX) const;
+
+        ///@}
+    };
+
+    class Data
```

Comment:
I'd suggest to find a more descriptive name for this class as we already have a bunch of data classes.

Reply:
Done 👍 . Changed to `QSVMSResidualData`